File: LinkedList/List.py

"""
PROGRAM TO CREATE A LINKED LIST IN PYTHON
"""

#   class used to create a single node or point to node
class Node : 
    data = None
    next_node = None


#   class to create a linked list
class LinkedList : 

    def __init__(self) : 
        self.head = None


    #   function to insert a node at the end of the linked list    
    def insertData(self, num) : 
        ptr = Node()
        ptr.data = num
        ptr.next_node = None
        
        if self.head == None : 
            self.head = ptr
        else : 
            qptr = self.head
            while qptr.next_node != None : 
                qptr = qptr.next_node

            qptr.next_node = ptr

    

    #   function to traverse the entire linked list (iterative method)
    def traverse(self) :
        if self.head == None : 
            print("No data in the list")
        else : 
            ptr = self.head
            while ptr != None : 
                print(ptr.data, end = " -> ") 
                ptr = ptr.next_node



    #   function to get the head pointer
    def gethead(self) :
        return self.head


    #   function to traverse a linked list recursively
    def recursiveTraverse(self, main_node) : 
        if main_node == None : 
            print("list ended")
            return

        print(main_node.data)
        self.recursiveTraverse(main_node.next_node)


    
    #   function to delete a node from the linked list using the data
    def remove(self, sk) : 
        if self.head == None : 
            print("The lis is empty")
            return

        if self.head.data == sk : 
            ptr = self.head
            self.head = ptr.next_node
            del ptr
            print("Element deleted")
            return

        qptr = self.head 
        
        while qptr.next_node != None : 

            if qptr.next_node.data == sk : 
                ptr = qptr.next_node
                qptr.next_node = ptr.next_node
                del ptr
                print("Element deleted ")
                return


            qptr = qptr.next_node

        print("Element was not found")       
        

    #   searching an element in the linked list
    def searchNode(self, sk) : 
        ptr = self.head
        if ptr is None : 
            print("List is empty");return

        while ptr is not None : 
            if ptr.data == sk : 
                print("Element was found in the list")
                return
            ptr = ptr.next_node

        print("Element was not found in the list")



    #   deleting the first node from the list
    def deleteFirst(self) : 
        if self.head is None : 
            print("list is empty")
            return

        ptr = self.head
        self.head = ptr.next_node
        num = ptr.data
        del ptr
        print("Deleted the first node")
        return num

    

    #   function to delete duplicate nodes from the linked list
    def delDuplicate(self) : 
        dup = set()
        if self.head is None : 
            print("List is empty")
            return

        ptr = self.head;prev=self.head

        dup.add(ptr.data)
        ptr = ptr.next_node

        while ptr is not None : 
            if ptr.data in dup : 
                prev.next_node = ptr.next_node
                del ptr
                ptr = prev.next_node
            else : 
                dup.add(ptr.data)
                prev = ptr
                ptr = ptr.next_node

            



if __name__ == '__main__' : 

    mylist = LinkedList()
    mylist.insertData(5)
    mylist.insertData(4)
    mylist.insertData(3)
    mylist.insertData(2)
    mylist.insertData(5)  

    mylist.traverse()

    print()
    mylist.delDuplicate()  

    mylist.traverse()

File: LinkedList/List.java

//  Data structure to create node of the linked list
class Node {
    int data;
    Node next;
}


class LinkedList {

    private Node head;

    //  constructor of the linked list class
    public LinkedList() {
        this.head = null;
    }

    //  method to append a node at the end of the linked list
    public void append(int sk) {

        Node ptr = new Node();
        ptr.data = sk;
        ptr.next = null;

        if (this.head == null) this.head=ptr;
        else {
            Node qptr = this.head;
            while (qptr.next != null) qptr=qptr.next;
            qptr.next = ptr;
        }

    }


    //  iterative method to traverse all the nodes of the linked list
    public void traverse() {

        Node ptr = this.head;
        if (ptr == null) {
            System.out.println("Linked list is empty");
            return;
        }
        while(ptr!=null) {
            System.out.println(ptr.data + " : ");
            ptr = ptr.next;
        }

    }

    //  method to return the head node of the linked list
    public Node getHead() { return this.head; }

    //  recursive method to traverse all the nodes of the linked list 
    public void recursiveTraversal(Node head) {

        if (head == null) {
            System.out.println("Printed all nodes");
            return;
        }

        System.out.print(head.data);
        this.recursiveTraversal(head.next);

    }

    //  method to count the number of nodes in the linked list
    public long countNodes() {
        Node ptr = this.head;
        long count = 0;
        while(ptr!=null) {
            count++;
            ptr=ptr.next;
        }
        return count;
    }

    //  method to calculate the sum of all the nodes in the linked list
    public long sum(Node head) {
        if (head == null) return 0;
        return head.data + sum(head.next);
    }

    //  method to get maximum
    public int getmax(int m, int n) {
        return m > n ? m : n;
    }

    //  method to get the maximum element from the linked list
    public int max() {

        if(this.head == null) return 0;

        int largest = this.head.data;
        Node ptr = this.head;

        while(ptr!=null) {
            largest = this.getmax(largest, ptr.data);ptr=ptr.next;
        }

        return largest;

    }

    public int search(int sk) {
        Node ptr = this.head;
        while(ptr!=null) {
            if(ptr.data == sk) {
                return ptr.data;
            }
            ptr = ptr.next;
        }
        return -1;
    }

    public boolean isAscending() {
        Node ptr = this.head;
        while(ptr.next!=null) {
            if(ptr.data > ptr.next.data) {
                return false;
            }
            ptr = ptr.next;
        }
        return true;
    }
    
}


class List {
    public static void main(String[] args) {
        LinkedList ob = new LinkedList();
        ob.append(1);
        ob.append(2);
        ob.append(3);
        ob.recursiveTraversal(ob.getHead());
        System.out.println("Total number of nodes = "+ob.countNodes());
        System.out.println("Sum of nodes of the list = "+ob.sum(ob.getHead()));
        System.out.println("Maximum element in the linked list = " + ob.max());
    }
}